### PR TITLE
WebGLRenderer: Avoid duplicate initialization of materials in compile()

### PR DESCRIPTION
**Pull request by EliasHasle**
Replaces #17116. Tested with the materials/compile example. Not tested for performance.

### CODE REVIEW

File: src/renderers/WebGLRenderer.js [hunk 1]
```diff
@@ -1055,6 +1055,8 @@ function WebGLRenderer( parameters ) {
 
 		currentRenderState.setupLights( camera );
 
+		let visited = {};
```
Comment (Mugen87) [inline src/renderers/WebGLRenderer.js]: Please use `var`  and a weak map. Otherwise you introduce a memory leak.

Comment (EliasHasle) [reply src/renderers/WebGLRenderer.js]: No memory leak. The object will be ready for garbage collection at the end of the function call, whether it is declared with a `var` or with a `let`. Note that I did specifically avoid trying to extend the memory of the visits beyond the `compile` call.

**Comment by EliasHasle** [comment]
I notice `initMaterial` takes a scene and an object as arguments in addition to the material. Isn't there a danger that this PR will cause errors by not calling `initMaterial` separately for multiple objects/scenes?

File: src/renderers/WebGLRenderer.js (hunk 1 above)
Comment (Mugen87) [reply src/renderers/WebGLRenderer.js]: Oh, of course. You're right! However, you should still use `var`.

**Comment by Mugen87** [comment]
Since `scene` is given as an argument, I don' think this should be no problem.

**Comment by EliasHasle** [comment]
`scene` is contant over the traversal, but `object` is not. Apparently, `WebGLPrograms` draws certain information from `object`, which may then not be properly considered if the respective `initMaterial` call is trimmed away.

File: src/renderers/WebGLRenderer.js (hunk 1 above)
Comment (EliasHasle) [reply src/renderers/WebGLRenderer.js]: I can use `var`. It makes no difference, but I guess it aids compatibility with Netscape Navigator. ;-)

**Comment by Mugen87** [comment]
This is a general problem of the engine since you can only have a single shader program per material anyway.

**Comment by EliasHasle** [comment]
You mean a new `initMaterial` with another object may potentially break the first?

**Comment by Mugen87** [comment]
Read #15047 and the related issues/PRs to understand the whole story.

**Comment by EliasHasle** [comment]
Thank you for the reference. If you follow the path outlined in #15047 any time soon, this PR will be obsolete. I will just await your decision, then.

**Comment by Mugen87** [comment]
I'm afraid this change won't be implemented soon. So this PR does have a purpose^^

**Review by Mugen87 — APPROVED** [review]

File: src/renderers/WebGLRenderer.js [hunk 2]
```diff
@@ -1055,6 +1055,8 @@ function WebGLRenderer( parameters ) {
 
 		currentRenderState.setupLights( camera );
 
+		var visited = {};
```
Comment (mrdoob) [inline src/renderers/WebGLRenderer.js]: Do you mind calling this `compiled`?

File: src/renderers/WebGLRenderer.js [hunk 3]
```diff
@@ -1063,13 +1065,19 @@ function WebGLRenderer( parameters ) {
 
 					for ( var i = 0; i < object.material.length; i ++ ) {
 
-						initMaterial( object.material[ i ], scene, object );
+						if ( ! visited[ object.material[ i ].uuid ] ) {
```
Comment (mrdoob) [inline src/renderers/WebGLRenderer.js]: `if ( object.material[ i ].uuid in compiled ) {`

File: src/renderers/WebGLRenderer.js [hunk 4]
```diff
@@ -1063,13 +1065,19 @@ function WebGLRenderer( parameters ) {
...
+
+							initMaterial( object.material[ i ], scene, object );
+							visited[ object.material[ i ].uuid ] = true;
+
+						}
 
 					}
 
-				} else {
+				} else if ( ! visited[ object.material.uuid ] ) {
```
Comment (mrdoob) [inline src/renderers/WebGLRenderer.js]: `} else if ( object.material.uuid in compiled ) {`

File: src/renderers/WebGLRenderer.js (hunk 3 above)
Comment (EliasHasle) [reply src/renderers/WebGLRenderer.js]: ...not. :wink:

**Comment by mrdoob** [comment]
Thanks!

**Comment by mrdoob** [comment]
> build/three.js:24547: WARNING - [JSC_SUSPICIOUS_NEGATED_LEFT_OPERAND_OF_IN_OPERATOR] Suspicious negated left operand of in operator.
					} else if ( ! object.material.uuid in compiled ) {
					            ^^^^^^^^^^^^^^^^^^^^^^

Actually, I don't think `if ( ! object.material[ i ].uuid in compiled === false ) {` works.

I think it should be `if ( object.material[ i ].uuid in compiled === false ) {` instead.

Good thin Closure Compiler caught that. Will fix.

**Comment by EliasHasle** [comment]
Yeah. I added `!` in front of your requested change, and must admit I was uncertain whether `!` would eat the whole `in` expression or not. It still worked on the compile example, so I kind of assumed it was OK, even though I knew the compile example would not test all cases. Sorry about that. Note that the PR is still _not_ thoroughly tested.

I was comfortable with my original expression, though. Isn't

```if ( ! compiled[ object.material[ i ].uuid ] )```

really better than

```if ( ! ( object.material[ i ].uuid ] in compiled ) )```

?

**Comment by mrdoob** [comment]
No worries!

I prefer explicit checks. Makes the code easier to follow.

**Pull request merged**